File: bridge/src/auth/signing-bytes.ts
/**
 * Canonical device-auth signing bytes and host normalization (spec §10.3).
 *
 * The signing content is the exact byte string both the Bridge and the
 * Android app construct for device-auth signatures:
 *
 * ```text
 * ASCII("CLAUDE-REMOTE-DEVICE-AUTH-V1") || 0x00 ||
 * u16be(len(hostAscii))       || UTF8(hostAscii) ||
 * u16be(len(deviceId))        || ASCII(deviceId) ||
 * u16be(len(challengeId))     || ASCII(challengeId) ||
 * u32be(len(accessSubject))   || UTF8(accessSubject) ||
 * challengeRaw[32]
 * ```
 *
 * All lengths are unsigned big-endian BYTE lengths of the following encoded
 * field. `accessSubject` is UTF-8 encoded as received — never Unicode
 * normalized. This module is pure: no I/O, no clock, no crypto.
 */
import { domainToASCII } from "node:url";

/** Domain-separation prefix of the signing content (§10.3). */
export const SIGNING_CONTEXT = "CLAUDE-REMOTE-DEVICE-AUTH-V1";

/** Thrown when a host/URL input cannot be normalized to a canonical host. */
export class InvalidHostException extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InvalidHostException";
  }
}

/** Thrown when the fields of the signing content are malformed. */
export class SigningInputError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SigningInputError";
  }
}

/**
 * Normalizes the Bridge public URL (or a bare canonical host) to the exact
 * `hostAscii` string used in signatures (§10.3):
 *
 * - scheme must be `https` (a bare hostname without scheme is also accepted,
 *   matching the stored canonical form);
 * - no userinfo, query, or fragment;
 * - path must be empty or `/`;
 * - port must be empty or `443` (WHATWG URL drops the default `:443`);
 * - hostname is IDNA ToASCII'd, lowercased, and stripped of its trailing dot;
 * - the scheme, path, and `:443` never appear in the signed host.
 */
export function normalizeHost(input: string): string {
  if (typeof input !== "string") {
    throw new InvalidHostException("host input must be a string");
  }
  const trimmed = input.trim();
  if (trimmed === "") {
    throw new InvalidHostException("host input is empty");
  }

  let hostname: string;
  if (/^[A-Za-z][A-Za-z0-9+.-]*:/.test(trimmed)) {
    let url: URL;
    try {
      url = new URL(trimmed);
    } catch {
      throw new InvalidHostException(`host input is not a valid URL: ${JSON.stringify(trimmed)}`);
    }
    if (url.protocol !== "https:") {
      throw new InvalidHostException(`host URL scheme must be https: ${JSON.stringify(trimmed)}`);
    }
    if (url.username !== "" || url.password !== "") {
      throw new InvalidHostException(`host URL must not contain userinfo: ${JSON.stringify(trimmed)}`);
    }
    if (url.search !== "") {
      throw new InvalidHostException(`host URL must not contain a query: ${JSON.stringify(trimmed)}`);
    }
    if (url.hash !== "") {
      throw new InvalidHostException(`host URL must not contain a fragment: ${JSON.stringify(trimmed)}`);
    }
    if (url.pathname !== "/") {
      throw new InvalidHostException(`host URL path must be empty or "/": ${JSON.stringify(trimmed)}`);
    }
    if (url.port !== "") {
      // url.port is "" for https even when the input spelled out ":443".
      throw new InvalidHostException(`host URL port must be empty or 443: ${JSON.stringify(trimmed)}`);
    }
    hostname = url.hostname;
  } else {
    if (/[/@?#:]/.test(trimmed)) {
      throw new InvalidHostException(`bare host must not contain URL delimiters: ${JSON.stringify(trimmed)}`);
    }
    hostname = trimmed;
  }

  const ascii = domainToASCII(hostname.toLowerCase());
  if (ascii === undefined || ascii === "") {
    throw new InvalidHostException(`hostname has no IDNA A-label form: ${JSON.stringify(hostname)}`);
  }
  const withoutRootDot = ascii.endsWith(".") ? ascii.slice(0, -1) : ascii;
  if (withoutRootDot === "") {
    throw new InvalidHostException("hostname is empty after normalization");
  }
  // domainToASCII passes structurally invalid ASCII input (e.g. "a..b")
  // through unchanged, so empty and non-LDH labels are rejected here.
  for (const label of withoutRootDot.split(".")) {
    if (!/^[a-z0-9]([a-z0-9-]*[a-z0-9])?$/.test(label)) {
      throw new InvalidHostException(
        `hostname contains an invalid label (${JSON.stringify(label)}) in ${JSON.stringify(input)}`,
      );
    }
  }
  if (withoutRootDot.length > 253) {
    throw new InvalidHostException("normalized hostname exceeds 253 characters");
  }
  return withoutRootDot;
}

/** Printable ASCII (no spaces, no control bytes) — required for host/device/challenge fields. */
const PRINTABLE_ASCII = /^[\x21-\x7e]+$/;

export interface SigningBytesInput {
  /** Canonical signed host (see {@link normalizeHost}). */
  readonly hostAscii: string;
  /** base64url_no_pad(SHA-256(SPKI DER)) — ASCII by construction. */
  readonly deviceId: string;
  /** Canonical lowercase UUID — ASCII by construction. */
  readonly challengeId: string;
  /** Verified Access subject string, UTF-8 encoded verbatim (no Unicode normalization). */
  readonly accessSubject: string;
  /** The 32-byte challenge; stored raw by the bridge for exact reconstruction. */
  readonly challengeRaw: Buffer;
}

function requirePrintableAscii(value: string, field: string): Buffer {
  if (typeof value !== "string" || value === "") {
    throw new SigningInputError(`${field} must be a non-empty string`);
  }
  if (!PRINTABLE_ASCII.test(value)) {
    throw new SigningInputError(`${field} must be printable ASCII`);
  }
  return Buffer.from(value, "ascii");
}

function u16be(value: number, field: string): Buffer {
  if (value > 0xffff) {
    throw new SigningInputError(`${field} exceeds the u16 length prefix`);
  }
  const buf = Buffer.alloc(2);
  buf.writeUInt16BE(value);
  return buf;
}

function u32be(value: number): Buffer {
  const buf = Buffer.alloc(4);
  buf.writeUInt32BE(value);
  return buf;
}

/**
 * Build the exact signing content for a device-auth signature. Throws
 * {@link SigningInputError} on any malformed field so callers never sign
 * degenerate input.
 */
export function buildSigningBytes(input: SigningBytesInput): Buffer {
  if (!Buffer.isBuffer(input.challengeRaw)) {
    throw new SigningInputError("challengeRaw must be a Buffer");
  }
  if (input.challengeRaw.length !== 32) {
    throw new SigningInputError(`challengeRaw must be exactly 32 bytes; got ${input.challengeRaw.length}`);
  }
  const hostBytes = requirePrintableAscii(input.hostAscii, "hostAscii");
  const deviceBytes = requirePrintableAscii(input.deviceId, "deviceId");
  const challengeIdBytes = requirePrintableAscii(input.challengeId, "challengeId");
  if (typeof input.accessSubject !== "string" || input.accessSubject === "") {
    throw new SigningInputError("accessSubject must be a non-empty string");
  }
  const subjectBytes = Buffer.from(input.accessSubject, "utf8");
  if (subjectBytes.length > 0xffffffff) {
    throw new SigningInputError("accessSubject exceeds the u32 length prefix");
  }

  return Buffer.concat([
    Buffer.from(SIGNING_CONTEXT, "ascii"),
    Buffer.from([0x00]),
    u16be(hostBytes.length, "hostAscii"),
    hostBytes,
    u16be(deviceBytes.length, "deviceId"),
    deviceBytes,
    u16be(challengeIdBytes.length, "challengeId"),
    challengeIdBytes,
    u32be(subjectBytes.length),
    subjectBytes,
    input.challengeRaw,
  ]);
}

File: android/app/src/main/java/dev/clauderemote/android/security/SigningBytes.kt
package dev.clauderemote.android.security

import java.net.IDN

/**
 * Canonical device-auth signing bytes and host normalization (spec §10.3).
 *
 * Byte-identical port of the reviewed TypeScript implementation
 * `bridge/src/auth/signing-bytes.ts`; the committed fixture
 * `contracts/v1/auth-signing-fixture.json` pins the exact layout and both
 * implementations' test suites assert against that same file.
 *
 * The signing content is the exact byte string both the Bridge and the
 * Android app construct for device-auth signatures:
 *
 * ```text
 * ASCII("CLAUDE-REMOTE-DEVICE-AUTH-V1") || 0x00 ||
 * u16be(len(hostAscii))       || UTF8(hostAscii) ||
 * u16be(len(deviceId))        || ASCII(deviceId) ||
 * u16be(len(challengeId))     || ASCII(challengeId) ||
 * u32be(len(accessSubject))   || UTF8(accessSubject) ||
 * challengeRaw[32]
 * ```
 *
 * All lengths are unsigned big-endian BYTE lengths of the following encoded
 * field. `accessSubject` is UTF-8 encoded as received — never Unicode
 * normalized. This file is pure JVM: no I/O, no clock, no crypto.
 */

/** Domain-separation prefix of the signing content (§10.3). */
const val SIGNING_CONTEXT = "CLAUDE-REMOTE-DEVICE-AUTH-V1"

/** Thrown when a host/URL input cannot be normalized to a canonical host. */
class InvalidHostException(message: String) : Exception(message)

/** Thrown when the fields of the signing content are malformed. */
class SigningInputError(message: String) : Exception(message)

/**
 * Normalizes the Bridge public URL (or a bare canonical host) to the exact
 * `hostAscii` string used in signatures (§10.3):
 *
 * - scheme must be `https` (a bare hostname without scheme is also accepted,
 *   matching the stored canonical form);
 * - no userinfo, query, or fragment;
 * - path must be empty or `/`;
 * - port must be empty or `443`;
 * - hostname is IDNA ToASCII'd, lowercased, and stripped of its trailing dot;
 * - the scheme, path, and `:443` never appear in the signed host.
 *
 * URL parsing mirrors the WHATWG parser the TypeScript implementation uses
 * for the checks above, but is deliberately stricter on two pathological
 * shapes WHATWG silently normalizes: percent-encoded host characters and
 * dot-segment paths (`/.`). Neither can produce a canonical host the two
 * implementations would agree on anyway, so they are rejected instead.
 */
fun normalizeHost(input: String): String {
    val trimmed = input.trim()
    if (trimmed.isEmpty()) {
        throw InvalidHostException("host input is empty")
    }

    val hostname = SCHEME_PREFIX.find(trimmed)?.let { match ->
        val scheme = match.value.dropLast(1).lowercase()
        if (scheme != "https") {
            throw InvalidHostException("host URL scheme must be https: \"$trimmed\"")
        }
        parseHttpsAuthority(trimmed.substring(match.value.length), trimmed)
    } ?: run {
        if (URL_DELIMITERS.containsMatchIn(trimmed)) {
            throw InvalidHostException("bare host must not contain URL delimiters: \"$trimmed\"")
        }
        trimmed
    }

    return toCanonicalAsciiHost(hostname, input)
}

/**
 * Parses everything after the `https:` scheme prefix and returns the raw
 * hostname (still mixed-case, possibly with a trailing root dot, possibly a
 * bracketed IP literal which later LDH validation rejects).
 */
private fun parseHttpsAuthority(afterScheme: String, originalInput: String): String {
    // WHATWG "special authority ignore slashes": for https any run of
    // forward/back slashes after the scheme is skipped before the authority.
    var authorityStart = 0
    while (authorityStart < afterScheme.length &&
        (afterScheme[authorityStart] == '/' || afterScheme[authorityStart] == '\\')
    ) {
        authorityStart++
    }
    val terminatorIdx = afterScheme.indexOfAny(AUTHORITY_TERMINATORS, authorityStart)
    val authorityEnd = if (terminatorIdx < 0) afterScheme.length else terminatorIdx
    val authority = afterScheme.substring(authorityStart, authorityEnd)

    // Userinfo runs up to the LAST '@' in the authority (WHATWG).
    if (authority.lastIndexOf('@') >= 0) {
        throw InvalidHostException("host URL must not contain userinfo: \"$originalInput\"")
    }

    // host[:port] — bracketed IP literals keep their brackets (rejected later
    // by LDH label validation, mirroring domainToASCII's undefined result).
    var host: String
    var portStr: String? = null
    if (authority.startsWith("[")) {
        val close = authority.indexOf(']')
        if (close < 0) {
            throw InvalidHostException("host input is not a valid URL: \"$originalInput\"")
        }
        host = authority.substring(0, close + 1)
        val afterBracket = authority.substring(close + 1)
        if (afterBracket.isNotEmpty()) {
            if (!afterBracket.startsWith(":")) {
                throw InvalidHostException("host input is not a valid URL: \"$originalInput\"")
            }
            portStr = afterBracket.substring(1)
        }
    } else {
        val colon = authority.indexOf(':')
        if (colon >= 0) {
            host = authority.substring(0, colon)
            portStr = authority.substring(colon + 1)
            if (portStr.contains(':')) {
                // Unbracketed multi-colon host (bare IPv6): not a URL WHATWG accepts.
                throw InvalidHostException("host input is not a valid URL: \"$originalInput\"")
            }
        } else {
            host = authority
        }
    }

    if (!portStr.isNullOrEmpty()) {
        val port = portStr.toIntOrNull()
        if (port == null || port !in 0..65535) {
            throw InvalidHostException("host input is not a valid URL: \"$originalInput\"")
        }
        if (port != 443) {
            throw InvalidHostException("host URL port must be empty or 443: \"$originalInput\"")
        }
    }

    // Path/query/fragment validation over the remainder (which starts at the
    // delimiter that terminated the authority, if any).
    val remainder = afterScheme.substring(authorityEnd)
    val hashIdx = remainder.indexOf('#')
    val beforeHash = if (hashIdx < 0) remainder else remainder.substring(0, hashIdx)
    val queryIdx = beforeHash.indexOf('?')
    val pathPart = if (queryIdx < 0) beforeHash else beforeHash.substring(0, queryIdx)
    val queryPart = if (queryIdx < 0) "" else beforeHash.substring(queryIdx + 1)
    val fragmentPart = if (hashIdx < 0) "" else remainder.substring(hashIdx + 1)

    // A lone '?' or '#' is dropped by WHATWG; only real content is rejected.
    if (pathPart.isNotEmpty() && pathPart != "/" && pathPart != "\\") {
        throw InvalidHostException("host URL path must be empty or \"/\": \"$originalInput\"")
    }
    if (queryPart.isNotEmpty()) {
        throw InvalidHostException("host URL must not contain a query: \"$originalInput\"")
    }
    if (fragmentPart.isNotEmpty()) {
        throw InvalidHostException("host URL must not contain a fragment: \"$originalInput\"")
    }

    return host
}

/**
 * IDNA ToASCII + lowercase + no trailing dot (§10.3), with the same LDH
 * label validation the TypeScript applies on top of `domainToASCII`.
 */
private fun toCanonicalAsciiHost(hostname: String, originalInput: String): String {
    val lower = hostname.lowercase()
    // Pure-ASCII hosts pass through WHATWG's non-strict domainToASCII
    // unchanged, so java.net.IDN is only consulted for non-ASCII input
    // (where it provides the IDNA A-label conversion). This also avoids
    // IDN.toASCII's stricter per-label length rejection of already-ASCII
    // labels, keeping acceptance aligned with the TypeScript.
    val ascii = if (lower.all { it.code < 0x80 }) {
        lower
    } else {
        try {
            IDN.toASCII(lower)
        } catch (e: IllegalArgumentException) {
            throw InvalidHostException("hostname has no IDNA A-label form: \"$hostname\"")
        }
    }
    val withoutRootDot = if (ascii.endsWith(".")) ascii.dropLast(1) else ascii
    if (withoutRootDot.isEmpty()) {
        throw InvalidHostException("hostname is empty after normalization")
    }
    // IDNA ToASCII passes structurally invalid ASCII input (e.g. "a..b")
    // through unchanged, so empty and non-LDH labels are rejected here.
    for (label in withoutRootDot.split('.')) {
        if (!LDH_LABEL.matches(label)) {
            throw InvalidHostException(
                "hostname contains an invalid label (\"$label\") in \"$originalInput\"",
            )
        }
        if (label.length > MAX_LABEL_LENGTH) {
            throw InvalidHostException("hostname label exceeds $MAX_LABEL_LENGTH characters")
        }
    }
    if (withoutRootDot.length > MAX_HOSTNAME_LENGTH) {
        throw InvalidHostException("normalized hostname exceeds $MAX_HOSTNAME_LENGTH characters")
    }
    return withoutRootDot
}

/** Printable ASCII (no spaces, no control bytes) — required for host/device/challenge fields. */
private val PRINTABLE_ASCII = Regex("^[\\x21-\\x7e]+$")

/** `scheme:` prefix — anything matching is URL-parsed rather than treated as a bare host. */
private val SCHEME_PREFIX = Regex("^[A-Za-z][A-Za-z0-9+.-]*:")

/** Characters a bare canonical host must never contain. */
private val URL_DELIMITERS = Regex("[/@#?:]")

/** LDH label: letters/digits/hyphens, no leading or trailing hyphen. */
private val LDH_LABEL = Regex("^[a-z0-9]([a-z0-9-]*[a-z0-9])?$")

/** Characters that terminate an authority component (backslash included: https is a special scheme). */
private val AUTHORITY_TERMINATORS = charArrayOf('/', '?', '#', '\\')

private const val MAX_LABEL_LENGTH = 63
private const val MAX_HOSTNAME_LENGTH = 253

/** Encodes a printable-ASCII field, rejecting empty or non-printable input. */
private fun printableAsciiBytes(value: String, field: String): ByteArray {
    if (value.isEmpty()) {
        throw SigningInputError("$field must be a non-empty string")
    }
    if (!PRINTABLE_ASCII.matches(value)) {
        throw SigningInputError("$field must be printable ASCII")
    }
    return value.toByteArray(Charsets.US_ASCII)
}

private fun u16be(value: Int, field: String): ByteArray {
    if (value > 0xffff) {
        throw SigningInputError("$field exceeds the u16 length prefix")
    }
    return byteArrayOf((value ushr 8).toByte(), value.toByte())
}

private fun u32be(value: Int): ByteArray =
    byteArrayOf(
        (value ushr 24).toByte(),
        (value ushr 16).toByte(),
        (value ushr 8).toByte(),
        value.toByte(),
    )

/**
 * Builds the exact signing content for a device-auth signature (§10.3).
 * Throws [SigningInputError] on any malformed field so callers never sign
 * degenerate input.
 *
 * @param hostAscii canonical signed host (see [normalizeHost]).
 * @param deviceId base64url_no_pad(SHA-256(SPKI DER)) — ASCII by construction.
 * @param challengeId canonical lowercase UUID — ASCII by construction.
 * @param accessSubject Verified Access subject string, UTF-8 encoded
 *   verbatim (no Unicode normalization).
 * @param challengeRaw the 32-byte challenge, stored raw by the bridge for
 *   exact reconstruction.
 */
fun buildSigningBytes(
    hostAscii: String,
    deviceId: String,
    challengeId: String,
    accessSubject: String,
    challengeRaw: ByteArray,
): ByteArray {
    if (challengeRaw.size != 32) {
        throw SigningInputError(
            "challengeRaw must be exactly 32 bytes; got ${challengeRaw.size}",
        )
    }
    val hostBytes = printableAsciiBytes(hostAscii, "hostAscii")
    val deviceBytes = printableAsciiBytes(deviceId, "deviceId")
    val challengeIdBytes = printableAsciiBytes(challengeId, "challengeId")
    if (accessSubject.isEmpty()) {
        throw SigningInputError("accessSubject must be a non-empty string")
    }
    val subjectBytes = accessSubject.toByteArray(Charsets.UTF_8)
    if (subjectBytes.size.toLong() > 0xffffffffL) {
        throw SigningInputError("accessSubject exceeds the u32 length prefix")
    }

    return SIGNING_CONTEXT.toByteArray(Charsets.US_ASCII) +
        byteArrayOf(0x00) +
        u16be(hostBytes.size, "hostAscii") + hostBytes +
        u16be(deviceBytes.size, "deviceId") + deviceBytes +
        u16be(challengeIdBytes.size, "challengeId") + challengeIdBytes +
        u32be(subjectBytes.size) + subjectBytes +
        challengeRaw
}
